lib/pymedphys/_experimental/streamlit/apps/sum_doses.py

```diff
@@ -0,0 +1,120 @@
+"""Copyright (C) 2021 Matthew Jennings
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import copy
+import pathlib
+
+from pymedphys._imports import streamlit as st
+
+import numpy as np
+
+import pydicom
+from pydicom.errors import InvalidDicomError
+
+from pymedphys._dicom.coords import xyz_axes_from_dataset
+from pymedphys._dicom.dose import dose_from_dataset
+from pymedphys._streamlit import categories
+
+STREAMLIT_STATIC_PATH = pathlib.Path(st.__path__[0]) / "static"
+DOWNLOADS_PATH = STREAMLIT_STATIC_PATH / "downloads"
+if not DOWNLOADS_PATH.is_dir():
+    DOWNLOADS_PATH.mkdir()
```

Comment:
Is this an issue? The standard way to provide a file to the user for download ([link](https://discuss.streamlit.io/t/how-to-download-file-in-streamlit/1806/9)) has a 50 MB write limit.

@sjswerdloff does seem to have a working solution to this in pseudonymise ([here](https://github.com/pymedphys/pymedphys/blob/ecd44331ec2dc06c04680b5ea03d4257b38e9d4c/lib/pymedphys/_streamlit/apps/pseudonymise.py#L84)). I had a bit of trouble adapting this solution for this app, and instead found this convenient fix ([associated link](https://github.com/streamlit/streamlit/issues/400#issuecomment-648580840)).

Happy to learn/adapt to a better fix. This was just nice and easy 😄

Reply:
Yeah, let's not use the `st.__path__[0]` file. Instead something like:

```
pathlib.Path.home() / ".pymedphys" / "downloads"
```

Might work well. At the end of the day, I'd expect this variable to be set within a config file and have some form of sensible default if it isn't in the file.

Reply:
Actually there is a better fourth option. Make a PR to streamlit that makes it so that there can be a configurable downloads path. (Or something equivalent that meets the need here).

I think I might looking to make that happen, that way, we can configure the downloads path to be within the user's drive, and it can be fully supported upstream (and useful to others with similar issues).

Reply:
> Actually there is a better fourth option. Make a PR to streamlit ...

I have found the best way forward with these sorts of things, is implement it first in PyMedPhys, show the working example, and then merge the important pieces of that upstream.

Reply:
As an initial work around so that this first pass can be merged, would you be able to nest this directory creation within a function? At the moment, if this was merged, the entire https://app.pymedphys.com may not work... as the "not allowed to make a directory" error will trigger on import (and therefore as the app boots). However, if you bring it in as a function it'll mean only your app won't work 🙂.

Reply:
> As an initial work around so that this first pass can be merged, would you be able to nest this directory creation within a function?

Done!

Reply:
> I have found the best way forward with these sorts of things, is implement it first in PyMedPhys, show the working example, and then merge the important pieces of that upstream.

I'm not sure I have the know-how to do this alone, but very keen to help/learn?

Reply:
> are you talking something as simple as a path in a config file?

Actually, after some more thought my idea is to make an `st_download(filename, path)` function to which a file path can be passed. Then the function will run off and make a specific URL point to that file directly, no matter where it is on the file system. The specific URL would have the session ID within it so as to have user's downloads not clash with each other, and a filename.

Began the work in the following PR:

https://github.com/pymedphys/pymedphys/pull/1388

Next part is to set up a thread-safe Queue (https://docs.python.org/3/library/queue.html), and then have a URL `http://localhost:8051/downloads/{session_id}/{filename}` that when called it checks to see if that session_id has run the `st_download` function with the defined filename. If it has, it points the URL to the filepath provided.

Reply:
Just a heads up. This was solved within PyMedPhys in https://github.com/pymedphys/pymedphys/pull/1402. Looking to merge upstream. Will now track within the following Streamlit issue:

https://github.com/streamlit/streamlit/issues/2722

Reply:
Yay! Should I wait and see if Streamlit merge in first?

Reply:
Nah, use our version. We can change it later. Currently, in your current implementation, if anyone installs PyMedPhys with a fresh install of Ubuntu 20.04 with the system version of Python, PyMedPhys and Streamlit will be installed with limited user file access rights. And if they click on your App, and have that function run, the Streamlit App will crash.

... Given PyMedPhys is a library that supports installation on a range of systems, leaving the download option as is, how it has been set up here isn't feasible.

And the version I have set up here should work just fine, even if it isn't merged upstream.